src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -1258,14 +1248,23 @@ BlankNode Visitor::visitTypesafe(Parser::BlankNodeContext* ctx) {
 
 // ____________________________________________________________________________________
 template <typename Out, typename FirstContext, typename... Context>
-Out SparqlQleverVisitor::visitAlternative(FirstContext ctx, Context... ctxs) {
+Out Visitor::visitAlternative(FirstContext ctx, Context... ctxs) {
   if (ctx)
     return {visitTypesafe(ctx)};
   else
     return visitAlternative<Out>(ctxs...);
 }
 
 template <typename Out>
-Out SparqlQleverVisitor::visitAlternative() {
+Out Visitor::visitAlternative() {
   AD_FAIL()  // Should be unreachable.
 }
+
+// ____________________________________________________________________________________
+template <typename Out, typename Ctx>
+std::optional<Out> Visitor::visitOptional(Ctx ctx) {
```

Comment:
In most cases the `Out` is redundant, because we simply want 
`Out = <result type of visitTypesafe(ctx)>`. So you can specify this as the default, 
you have already mastered the shenanigans around decltype and declval.

If we never want a different type, you can get away with
```
template <typename Ctx>
auto Visitor::visitOptional(Ctx ctx) -> std::optional<decltype(visitTypesafe(ctx))> {
...
}
```

And I think this suffices here... (we can still change it, if we want the conversion).